performance_manager/lib/rt_trip_updates.py

```diff
@@ -0,0 +1,250 @@
+        "route_id": row["route_id"],
+        "start_date": row["start_date"],
+        "start_time": row["start_time"],
+        "vehicle_id": row["vehicle_id"],
+    }
+    return_list: List[Dict[str, Any]] = []
+    for record in row["stop_time_update"]:
+        try:
+            arrival_time = int(record["arrival"]["time"])
+        except (TypeError, KeyError):
+            continue
+        if (
+            arrival_time - row["timestamp"] < 0
+            or arrival_time - row["timestamp"] > 120
+        ):
+            continue
+        append_dict.update(
+            {
+                "stop_id": record.get("stop_id"),
+                "stop_sequence": record.get("stop_sequence"),
+                "timestamp_start": arrival_time,
+            }
+        )
+        return_list.append(append_dict.copy())
+
+    if len(return_list) == 0:
+        return None
+
+    return return_list
+
+
+def unwrap_tu_dataframe(events: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    unwrap and explode trip updates records from parquet files
+    parquet files contain stop_time_update field that is saved as list of dicts
+    stop_time_update must have fields extracted and flattened to create
+    predicted trip update stop events
+    """
+
+    # store start_date as int64
+    events["start_date"] = pandas.to_numeric(events["start_date"]).astype(
+        "int64"
+    )
+
+    # store direction_id as int64
+    events["direction_id"] = pandas.to_numeric(events["direction_id"]).astype(
+        "int64"
+    )
+
+    # store start_time as seconds from start of day int64
+    events["start_time"] = (
+        events["start_time"].apply(start_time_to_seconds).astype("int64")
+    )
+
+    # expand and filter stop_time_update column
+    # this will return a Series with values being list of dicts
+    events = events.apply(explode_stop_time_update, axis=1).dropna()
+
+    # transform Series of list of dicts into dataframe
+    events = pandas.json_normalize(events.explode())
+
+    # is_moving column to indicate all stop events
+    # needed for later join with vehicle_position_event by hash
+    events["is_moving"] = False
+    events["pk_id"] = None
+
+    # add hash column, hash should be consistent across trip_update and
+    # vehicle_position events
+    events = add_event_hash_column(events).sort_values(by=["hash", "timestamp"])
+
+    # after sort, drop all duplicates by hash, keep last record
+    # last record will be most recent arrival time prediction for event
+    events = events.drop_duplicates(subset=["hash"], keep="last")
+
+    # after sort and drop timestamp column no longer needed
+    events = events.drop(columns=["timestamp"])
+
+    return events
+
+
+def merge_trip_update_events(
+    new_events: pandas.DataFrame, session: sa.orm.session.sessionmaker
+) -> None:
+    """
+    merge new trip update evetns with existing events found in database
+    merge performed on hash of records
+    """
```

Comment:
This is also very similar to the `RT_VEHICLE_POSITIONS` logic, however these events only have `timestamp_start` no `timestamp_end` and they are all "stopped_at" events.

Reply:
i think its fine to keep these as different flows. if we end up having a third that has commonality, it might be worth doing some abstraction, but this all reads clear enough to me.